Enable ARM's LSE (Large System Extensions) for atomic operations

davidzengxhsh opened the pull request:
LSE can reduce the time of Q30 of the hits dataset from 2.7s to 0.5s.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

davidzengxhsh committed:
Check the LSE support on AArch64 and add "+lse" for performance.
It can reduce the time of Q30 of the hits dataset from 2.7s to 0,5s.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=40509) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=40509) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=40509) it.</sub>

rschu1ze commented:
@davidzengxhsh Would you like to sign the CLA first? I will start builds/tests afterwards.

davidzengxhsh commented:
> @davidzengxhsh Would you like to sign the CLA first? I will start builds/tests afterwards.

done and thanks!

rschu1ze commented on cmake/cpu_features.cmake:
```diff
@@ -30,7 +30,13 @@ if (ARCH_NATIVE)
 
 elseif (ARCH_AARCH64)
-    set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=armv8-a+crc+simd+crypto+dotprod+ssbs")
+    CHECK_C_COMPILER_FLAG("-march=armv8-a+lse" HAS_ARMV8_LSE)
```
`CHECK_C_COMPILER_FLAG` doesn't compile anything, it merely checks if the C compiler accepts the passed arguments. A ClickHouse build requires minimum versions of Clang and GCC, see "cmake/tools.cmake" for that, and these versions are rather new. We can simply assume that the compiler knows the `lse` flag.

davidzengxhsh replied:
Yes, that make sense. Let me remove the checking and just add the "lse" into the feature list.

rschu1ze commented:
So just to explain the background: LSE is the abbreviation for "Large System Extensions" and it it ARM's optimized instruction set extension for atomic operations. It was made mandatory with ARM v8.1 which was released in 2015 and the first processors implementing ARM v8.1 appeared in 2016. For reference, ARM v8.0 was released in 2011 and the first (consumer-grade) processors implementing v8.0 appeared in 2013. In particular, the first generation of Amazon Graviton processors does not implement v8.1 (it implements v8.0).

I bring up all these dates because
a) on x86, there was a lot of [discussion](https://github.com/ClickHouse/ClickHouse/issues/40459) if ClickHouse can assume availability of AVX 1 instructions which were released in 2011. There was a backlash from the embedded/hobbyist community who run ClickHouse on older (or recent) hardware w/o AVX but also from users maintaining server in which some systems are rather old.
b) ClickHouse will soon launch a cloud offering on top of AWS. I'll ask we need to stay downwards compatible with 1st gen Gravitons. But I could imagine similar issues for other companies hosting ClickHouse instances.

To mitigate the compat issues, GCC offers `-moutline-atomics` (see [here](https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/making-the-most-of-the-arm-architecture-in-gcc-10), [here](https://mysqlonarm.github.io/ARM-LSE-and-MySQL/) and [here](https://dev.to/aws-builders/large-system-extensions-for-aws-graviton-processors-3eci)) which does a runtime-dispatch to non-atomic or LSE instructions depending on the target hardware. And it seems they made that switch default with GCC >= 10. The problem is that ClickHouse is compiled with Clang (GCC works theoretically too but is currently broken and thus disabled). So it would be useful to find out if the switch is default-enabled on [Clang](https://clang.llvm.org/docs/ClangCommandLineReference.html) too and if yes starting from which version. I would say only has a realistic chance of getting merged with the switch turned on.

Let's wait for performance results and check the improvements.

davidzengxhsh committed:
Since ClickHouse build has minimum GCC/Clang version requirement, remove the checking and add "lse" option directly.

rschu1ze commented on cmake/cpu_features.cmake:
```diff
@@ -30,7 +30,7 @@ if (ARCH_NATIVE)
 
 elseif (ARCH_AARCH64)
-    set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=armv8-a+crc+simd+crypto+dotprod+ssbs")
+    set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=armv8-a+crc+simd+crypto+dotprod+ssbs+lse")
```
As I don't have permissions to push to your fork, would you like to add the following lines below l. 33?

```cmake
# The LSE extension for atomic operations is mandatory in ARM >= v8.1. As we target v8.0,
# use a runtime-dispatch that falls back to non-atomic operations hardware w/o LSE.
set (COMPILER_FLAGS "${COMPILER_FLAGS} -moutline-atomics")
```

davidzengxhsh replied:
The -moutline-atomics was supported by clang-12(see [here](https://reviews.llvm.org/rG4d7df43ffdb460dddb2877a886f75f45c3fee188)) and gcc 10.1. The minimal compiler requirement of ClickHouse(gcc 11 or clang 12) contains this support. 
Compared to +lse, -moutline-atomics provides better compatibility with a little bit performance impact.

davidzengxhsh committed:
Use -moutline-atomics for AArch64 LSE support.
ClickHouse defines the minimal compiler version gcc11 and clang12, both support this option.

rschu1ze commented on cmake/cpu_features.cmake:
```diff
@@ -30,7 +30,13 @@ if (ARCH_NATIVE)
+    set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=armv8-a+crc+simd+crypto+dotprod+ssbs")
+
+    # The LSE extension for atomic operations is mandatory in ARM >= v8.1. As we target v8.0,
+    # use a runtime-dispatch that falls back to non-atomic operations hardware w/o LSE.
```
```suggestion
    # use a runtime-dispatch that falls back to non-atomic operations on hardware w/o LSE.
```

rschu1ze commented:
With `-moutline-atomics`, the ARM build fail to link:
```
Sep 07 06:28:14 FAILED: utils/self-extracting-executable/decompressor 
Sep 07 06:28:14 : && /usr/bin/clang++-14 --target=aarch64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-aarch64/aarch64-linux-gnu/libc  --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-aarch64 --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-aarch64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -march=armv8-a+crc+simd+crypto+dotprod+ssbs -moutline-atomics -fasynchronous-unwind-tables -ffile-prefix-map=/build=. -falign-functions=32 -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -O2 -g -DNDEBUG -O3 -g -gdwarf-4  --ld-path=/usr/bin/ld.lld-14 -rdynamic -Wl,--gdb-index -Wl,--build-id=sha1    -Xlinker --no-undefined utils/self-extracting-executable/CMakeFiles/decompressor.dir/decompressor.cpp.o  -o utils/self-extracting-executable/decompressor  contrib/zstd-cmake/lib_zstd.a  -Wl,--start-group  base/glibc-compatibility/libglibc-compatibility.a  contrib/libcxx-cmake/libcxx.a  contrib/libcxxabi-cmake/libcxxabi.a  contrib/libunwind-cmake/libunwind.a  -Wl,--end-group  -nodefaultlibs -lgcc  -lc -lm -lrt -lpthread -ldl && :
Sep 07 06:28:14 ld.lld-14: error: undefined symbol: __aarch64_ldadd8_acq_rel
Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::filesystem_error::~filesystem_error()) in archive contrib/libcxx-cmake/libcxx.a
Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::filesystem_error::~filesystem_error()) in archive contrib/libcxx-cmake/libcxx.a
Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::__fs_is_empty(std::__1::__fs::filesystem::path const&, std::__1::error_code*)) in archive contrib/libcxx-cmake/libcxx.a
Sep 07 06:28:14 >>> referenced 317 more times
```

I believe this is related to https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1802783.html

davidzengxhsh commented:
> With `-moutline-atomics`, the ARM build fail to link:
> 
> ```
> Sep 07 06:28:14 FAILED: utils/self-extracting-executable/decompressor 
> Sep 07 06:28:14 : && /usr/bin/clang++-14 --target=aarch64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-aarch64/aarch64-linux-gnu/libc  --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-aarch64 --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-aarch64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -march=armv8-a+crc+simd+crypto+dotprod+ssbs -moutline-atomics -fasynchronous-unwind-tables -ffile-prefix-map=/build=. -falign-functions=32 -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -O2 -g -DNDEBUG -O3 -g -gdwarf-4  --ld-path=/usr/bin/ld.lld-14 -rdynamic -Wl,--gdb-index -Wl,--build-id=sha1    -Xlinker --no-undefined utils/self-extracting-executable/CMakeFiles/decompressor.dir/decompressor.cpp.o  -o utils/self-extracting-executable/decompressor  contrib/zstd-cmake/lib_zstd.a  -Wl,--start-group  base/glibc-compatibility/libglibc-compatibility.a  contrib/libcxx-cmake/libcxx.a  contrib/libcxxabi-cmake/libcxxabi.a  contrib/libunwind-cmake/libunwind.a  -Wl,--end-group  -nodefaultlibs -lgcc  -lc -lm -lrt -lpthread -ldl && :
> Sep 07 06:28:14 ld.lld-14: error: undefined symbol: __aarch64_ldadd8_acq_rel
> Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
> Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::filesystem_error::~filesystem_error()) in archive contrib/libcxx-cmake/libcxx.a
> Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
> Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::filesystem_error::~filesystem_error()) in archive contrib/libcxx-cmake/libcxx.a
> Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
> Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::__fs_is_empty(std::__1::__fs::filesystem::path const&, std::__1::error_code*)) in archive contrib/libcxx-cmake/libcxx.a
> Sep 07 06:28:14 >>> referenced 317 more times
> ```
> 
> I believe this is related to https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1802783.html

I can get through with -moutline-atomics on a native AArch64 server with CentOS 8.5 using following command:
cmake .. -DENABLE_AZURE_BLOB_STORAGE=0 -DENABLE_S3=0 -DENABLE_AVRO=0 -DENABLE_BASE64=0 -DENABLE_CCACHE=0
ninja clickhouse

And I checked the generated ./contrib/libcxx-cmake/libcxx.a, it contains
# nm ./contrib/libcxx-cmake/libcxx.a |grep aarch64
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd4_rel
                 U __aarch64_cas1_acq_rel
                 U __aarch64_swp8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_relax
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_relax
                 U __aarch64_cas8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_relax

Are you using a cross compiler?

davidzengxhsh commented:
> With `-moutline-atomics`, the ARM build fail to link:
> 
> ```
> Sep 07 06:28:14 FAILED: utils/self-extracting-executable/decompressor 
> Sep 07 06:28:14 : && /usr/bin/clang++-14 --target=aarch64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-aarch64/aarch64-linux-gnu/libc  --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-aarch64 --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-aarch64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -march=armv8-a+crc+simd+crypto+dotprod+ssbs -moutline-atomics -fasynchronous-unwind-tables -ffile-prefix-map=/build=. -falign-functions=32 -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -O2 -g -DNDEBUG -O3 -g -gdwarf-4  --ld-path=/usr/bin/ld.lld-14 -rdynamic -Wl,--gdb-index -Wl,--build-id=sha1    -Xlinker --no-undefined utils/self-extracting-executable/CMakeFiles/decompressor.dir/decompressor.cpp.o  -o utils/self-extracting-executable/decompressor  contrib/zstd-cmake/lib_zstd.a  -Wl,--start-group  base/glibc-compatibility/libglibc-compatibility.a  contrib/libcxx-cmake/libcxx.a  contrib/libcxxabi-cmake/libcxxabi.a  contrib/libunwind-cmake/libunwind.a  -Wl,--end-group  -nodefaultlibs -lgcc  -lc -lm -lrt -lpthread -ldl && :
> Sep 07 06:28:14 ld.lld-14: error: undefined symbol: __aarch64_ldadd8_acq_rel
> Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
> Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::filesystem_error::~filesystem_error()) in archive contrib/libcxx-cmake/libcxx.a
> Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
> Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::filesystem_error::~filesystem_error()) in archive contrib/libcxx-cmake/libcxx.a
> Sep 07 06:28:14 >>> referenced by shared_ptr.h:118 (../contrib/libcxx/include/__memory/shared_ptr.h:118)
> Sep 07 06:28:14 >>>               operations.cpp.o:(std::__1::__fs::filesystem::__fs_is_empty(std::__1::__fs::filesystem::path const&, std::__1::error_code*)) in archive contrib/libcxx-cmake/libcxx.a
> Sep 07 06:28:14 >>> referenced 317 more times
> ```
> 
> I believe this is related to https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1802783.html

I can get through with -moutline-atomics on a native AArch64 server with CentOS 8.5 using following command:
cmake .. -DENABLE_AZURE_BLOB_STORAGE=0 -DENABLE_S3=0 -DENABLE_AVRO=0 -DENABLE_BASE64=0 -DENABLE_CCACHE=0
ninja clickhouse

And I checked the generated ./contrib/libcxx-cmake/libcxx.a, it contains
# nm ./contrib/libcxx-cmake/libcxx.a |grep aarch64
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd4_rel
                 U __aarch64_cas1_acq_rel
                 U __aarch64_swp8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_relax
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_relax
                 U __aarch64_cas8_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd4_acq_rel
                 U __aarch64_ldadd8_acq_rel
                 U __aarch64_ldadd8_relax

Are you using a cross compiler?

davidzengxhsh committed:
Update cmake/cpu_features.cmake

Co-authored-by: Robert Schulze <robert@clickhouse.com>

davidzengxhsh committed:
Update cmake/cpu_features.cmake

Co-authored-by: Robert Schulze <robert@clickhouse.com>

davidzengxhsh commented:
> 
I think the feature get from /proc/cpuinfo is _atomics_, not _lse_:
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 **atomics** fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs

rschu1ze committed:
Update cmake/cpu_features.cmake

rschu1ze committed:
Merge branch 'master' into master

rschu1ze commented:
Update:

- Native compilation on Aarch64 for Aarch64 works and the produced binaries run fine. The only modification over the standard build was to add `set (COMPILER_FLAGS "${COMPILER_FLAGS} -moutline-atomics")` to "cpu_features.cmake". This works independently of `-DENABLE_CLICKHOUSE_SELF_EXTRACTING=` `0` or `1`. My local CPU had no LSE, i.e. it was strictly ARMv8.0.

- The CI builds for Aarch64 are based on cross-compilation on x86 as host system.
  1. If one goes for an earlier version of this PR (first commit) where only `+lse` is added to the Aarch64 compile flags in "cpu_features.cmake", everything is fine and this is also where we saw performance wins. Guess the CI test machines support LSE. Didn't try this variant locally.
  2. If one goes for the current version where just `set (COMPILER_FLAGS "${COMPILER_FLAGS} -moutline-atomics")` is added, *all* binaries fail to link due to unresolved symbols. "pre_compressor" is just the binary which happens to be linked first (it is a little bit confusing because "pre_compressor" is build twice in cross-builds, once for the host platform x86 (works) and once for the target platform aarch64 (fails)). To reproduce:
  ```
  # add `set (COMPILER_FLAGS "${COMPILER_FLAGS} -moutline-atomics")` below `set (COMPILER_FLAGS "${COMPILER_FLAGS} -march=armv8-a+crc+simd+crypto+dotprod+ssbs")` to cpu_features.cmake
  $ cmake -DCMAKE_BUILD_TYPE=None -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_TOOLCHAIN_FILE=cmake/linux/toolchain-aarch64.cmake -S . -B build
  $ cd build
  $ ninja utils/config-processor/config-processor
  ```
  moutline-atomics calls a runtime dispatch helper which is part of the C library (see [here](https://dev.to/aws-builders/large-system-extensions-for-aws-graviton-processors-3eci), "Is LSE built into the C library on my operating system?"). As far as I see using `nm`, the libraries in contrib/sysroot/linux-aarch64/ have neither of the [missing symbols](https://github.com/ClickHouse/ClickHouse/pull/40509#issuecomment-1240177595). Chances are that the libs are too old (e.g. pre-2020 when GCC 10.1 introduced moutline-atomics).

@amosbird @alexey-milovidov Tldr; This PR enables LSE on ARM, i.e. faster atomics, but it needs an sysroot update for Aarch64. Do you remember how the sysroot was created? E.g. just copypaste the libs from a recent Ubuntu?

alexey-milovidov commented:
We may need to add two AArch64 builds: one for the most modern CPUs like Graviron3 and what's in Google Cloud, and another build for Raspberry Pi and similar. See #41155.

> it needs an sysroot update for Aarch64

They were downloaded from developer.arm.com, see https://github.com/ClickHouse/sysroot/#readme
But you can simply get the Docker image for Ubuntu and copy the relevant files. Or you can use the ISO image if you like...

By the way, will it make our build incompatible with old glibc? It will be not good...
We have compatibility for anything as old as Ubuntu 8.04 on x64_64, and I think we can start from Ubuntu 20.04 for aarch64.

rschu1ze commented:
> We may need to add two AArch64 builds: one for the most modern CPUs like Graviron3 and what's in Google Cloud, and another build for Raspberry Pi and similar. See #41155.

Interesting, I expected that ClickHouse works on Raspberry Pi 4. It has a 2018 ARM A-72 CPU with ARM-v8.0 support. This is the ISA level we compile for. But we also add flags `crc+simd+crypto+dotprod+ssbs` which I guess the Raspberry doesn't support

I like the idea to provide a "basic-level" ARM build (similar to the SSE2 build) and a "cloud-level" build. For the latter, we can add the LSE flag directly w/o `-moutline-atomics`. Will ask internally what the baseline is for CH Cloud.

> > it needs an sysroot update for Aarch64
> They were downloaded from developer.arm.com, see https://github.com/ClickHouse/sysroot/#readme But you can simply get the Docker image for Ubuntu and copy the relevant files. Or you can use the ISO image if you like...

Thanks. The upgrade hassle is only necessary for `-moutline-atomics`. I think we can avoid that.

> By the way, will it make our build incompatible with old glibc? It will be not good... We have compatibility for anything as old as Ubuntu 8.04 on x64_64, and I think we can start from Ubuntu 20.04 for aarch64.

`-moutline-atomics` would break downwards compat but Ubuntu 20.04 would be recent enough. But with two builds that's not relevant anyways.

rschu1ze commented:
@davidzengxhsh I am continuing the idea of this PR in #41610. Feel free to close this PR. Thanks for your contribution.

davidzengxhsh commented:
> 

Sure. Thanks!
